### PR TITLE
[Merged by Bors] - Improve soundness of `CommandQueue`

### PR DESCRIPTION
**Pull request by SkiFire13**
# Objective

This PR aims to improve the soundness of `CommandQueue`. In particular it aims to:
- make it sound to store commands that contain padding or uninitialized bytes;
- avoid uses of commands after moving them in the queue's buffer (`std::mem::forget` is technically a use of its argument);
- remove useless checks: `self.bytes.as_mut_ptr().is_null()` is always `false` because even `Vec`s that haven't allocated use a dangling pointer. Moreover the same pointer was used to write the command, so it ought to be valid for reads if it was for writes.

## Solution

- To soundly store padding or uninitialized bytes `CommandQueue` was changed to contain a `Vec<MaybeUninit<u8>>` instead of `Vec<u8>`;
- To avoid uses of the command through `std::mem::forget`, `ManuallyDrop` was used.
 
## Other observations

While writing this PR I noticed that `CommandQueue` doesn't seem to drop the commands that weren't applied. While this is a pretty niche case (you would have to be manually using `CommandQueue`/`std::mem::swap`ping one), I wonder if it should be documented anyway.

### CODE REVIEW

File: crates/bevy_ecs/src/system/commands/command_queue.rs [hunk 1]
```diff
@@ -81,27 +85,11 @@ impl CommandQueue {
         // unnecessary allocations.
         unsafe { self.bytes.set_len(0) };
 
-        let byte_ptr = if self.bytes.as_mut_ptr().is_null() {
```
Comment (TheRawMeatball) [inline crates/bevy_ecs/src/system/commands/command_queue.rs:84]: Why is this removed?

**Review by TheRawMeatball — APPROVED** [review]
The changes look good, but the justification for removing the null check should be added as a comment.

**Comment by BoxyUwU** [comment]
Can you add a test that miri would yell about for the uninit memory case

**Review by DJMcNab — APPROVED** [review]

File: crates/bevy_ecs/src/system/commands/command_queue.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (SkiFire13) [reply crates/bevy_ecs/src/system/commands/command_queue.rs:84]: Going by what the docs say I don't think we can program against them, `as_mut_ptr` could also return a freed pointer, which according to https://doc.rust-lang.org/stable/std/ptr/index.html#safety are also not valid for zero-sized reads. I guess the correct way to do this would be checking if the capacity is 0 rather than if the pointer is null. Anyway, I opened a PR to change `Vec::as_ptr` and `Vec::as_mut_ptr`'s docs, so hopefully this get clarified in the stdlib.

**Comment by alice-i-cecile** [comment]
> (std::mem::forget is technically a );
I think you missed part of the sentence here.

**Comment by SkiFire13** [comment]
> > (std::mem::forget is technically a );
> I think you missed part of the sentence here.

Oh right, I was trying to rephrase it and then suddently `std::mem::forget` about it. Anyway, I fixed it.
